staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -916,6 +917,24 @@ func (r *Request) newHTTPRequest(ctx context.Context) (*http.Request, error) {
 	return req, nil
 }
 
+// newDNSMetricsTrace returns an HTTP trace that tracks time spend on DNS lookups per network/host.
+// This metric is available in client as "rest_client_resolver_duration_seconds".
+func newDNSMetricsTrace(ctx context.Context) *httptrace.ClientTrace {
+	var (
+		dnsStart time.Time
+		dnsHost  string
```

Comment:
last time I checked this with Jordan I think we concluded this is not thread safe, so you have to lock the access to this vars

Reply:
@aojea interesting, even if they are local to that function?

Reply:
I vaguely remember that is because the request can be retried, talking from memory now

Reply:
we only retry a request from a single thread, but I don't know enough about stdlib behavior to know if dns resolution ever happens in parallel... I would assume it's possible

Reply:
I think that this is ok

Reply:
can we find a definitive answer in the stdlib tracing docs about whether these are invoked in parallel? I don't want to accidentally introduce a data race based on a point-in-time code inspection

Reply:
I was able to reproduce the data race in a test, then I re-read the docs for httptrace and noticed this:

> ClientTrace is a set of hooks to run at various stages of an outgoing HTTP request. Any particular hook may be nil. **Functions may be called concurrently from different goroutines** and some may be called after the request has completed or failed.

It seems that the DNS request themselves are coalesced, but the traces are intentionally not. Also came across this related TODO, removed during the implementation of tracing:

> when adding trace hooks later here, make sure the tracing is done outside of the singleflight merging. Both callers should see the DNS lookup delay, even if it's only being done once.

https://github.com/golang/go/blob/b6b4004d5a5bf7099ac9ab76777797236da7fe63/src/net/lookup.go#L90C1-L94

Reply:
> I was able to reproduce the data race in a test

![image](https://github.com/kubernetes/kubernetes/assets/980082/646dc0a2-0309-43ce-b4b9-b780c407fd5d)